### ISSUE TITLE
Ability to add/replace shader and material properties for bevy_sprite

### ISSUE THREAD

**Issue by recatek** [bevyengine/bevy#1738]
## What problem does this solve or what need does it fill?
The render pipeline for bevy_sprite is currently a bit rigid. In order to take advantage of the sprite sheet and other nice features of the crate, you currently need to commit to a hardcoded vertex and fragment shader. This limits your ability to control how your sprites are drawn. One use case of custom shaders on sprites is to [dynamically recolor masked portions of the sprite for team colors](https://bgolus.medium.com/the-team-color-problem-b70ec69d109f). To achieve this right now you need to dig pretty deep into the sprite code to see how it builds its pipeline and render graph nodes, and either replace them or clone the crate with your own changes.

## What solution would you like?

Ideally, a more surface-level way to customize/override the shader and material that the sprite system uses, while preserving its functionality for sprite sheets and other nice-to-haves, like sizing.

## What alternative(s) have you considered?

You can fork bevy_sprite and use your own implementation, or reimplement portions of it piecemeal (like `build_sprite_pipeline`). You can override the `material` and the `render_pipelines` on the sprite entity with a custom bundle, but that severs you from the nice sprite functionality like automatic resizing and doesn't seem to be better than spawning a quad and going out on your own. See [example gist](https://gist.github.com/recatek/58f60b04521ef99ec39ccbe32cc5b0c3).

## Additional context

Reconciling user-custom shaders and materials with the requirements of how the sprite sheet system works will be tricky. I'm not sure how to do this at first glance other than clearly documenting what the user-provided shader needs to support and how in order to play nicely with sprite sheets.

**Comment by vsaase** [comment]
not even the gist (after some slight modifications to make it compile) seems to work with bevy 0.5 now:  no error, no sprite

**Comment by jakudlaty** [comment]
Any chance this issue will get some attention? :) It would be really nice to have a convinient way to use shaders with TextureAtlasSprite and Sprite. Unfortunatelly im to new to Bevy and rust to create a PR for that :(

**Comment by FelixSelter** [comment]
Agreed. We need a way to do this. Currently there's [this](https://rustrepo.com/repo/ManevilleF-bevy_sprite_material)
But it requires bevy 7

**Comment by MatrixDev** [comment]
It would be really nice to have this feature. Without it sprites are very limited... :(

**Comment by ElliotB256** [comment]
What about mirroring the approach used for `CustomMaterial`, and modify `SpritePipeline` into a `pub struct SpritePipeline<M : SpriteMaterial>`, with logic for setting up the render state and shader moved into methods on the `SpriteMaterial: AsBindGroup + Asset + Clone + Sized`?

**Comment by ElliotB256** [comment]
There's some additional complexity required to maintain functionality of the batcher. In the current sprite pipeline, the only `group(1)` binding that could vary from one sprite to the next is the texture, so the sprite pipeline compares whether the next sprite to be drawn is equivalent to the previous in binding. If they are, they are combined into a batch, if they aren't, a new batch is created.

A custom `SpriteMaterial` trait could provide a method to compare if two sprites can be batched together, which would allow batching to be supported. A straightforward example would be a base texture and a team color mask. However, there's probably going to be a bunch of properties that are almost unbatchable - for example, colors/floats used for 'damage flash' effects in arcade games. What would be the most performant way to handle these?

**Comment by leomeinel** [comment]
I'm pretty sure this can be closed since https://github.com/bevyengine/bevy/pull/22484 has been merged, but I'm not sure if it will provide everything that has been asked for.
